File: everyDayGo/808.分汤.go
/*
 * @lc app=leetcode.cn id=808 lang=golang
 *
 * [808] 分汤
 */

// @lc code=start
func soupServings(n int) float64 {
	n = (n + 24) / 25
	if (n >= 200) {
		n = 200
	}
    dp := make([][]float64, n+1)
    for i := range dp {
        dp[i] = make([]float64, n+1)
    }
	dp[0][0] = 0.5
	for i := 1; i <= n; i++ {
		dp[0][i] = 1.0;
		dp[i][0] = 0;
	}
	for i := 1; i <= n; i++ {
		for j := 1; j<= n; j++ {
			dp[i][j] = (dp[max(i - 4, 0)][j] + dp[max(i - 3, 0)][j - 1] + dp[max(i - 2, 0)][max(j - 2, 0)] + dp[i - 1][max(j - 3, 0)]) / 4.0
		}
	} 
 	return dp[n][n]
}

func max(a int, b int) int {
	if (a > b) {
		return a
	}
	return b
}
// @lc code=end

File: everyDay/808.分汤.cpp
/*
 * @lc app=leetcode.cn id=808 lang=cpp
 *
 * [808] 分汤
 */

// @lc code=start
class Solution {
public:
    double soupServings(int n) {
        n = min(int(ceil((double) n / 25)), 200);
        // double dp[n + 4][n + 4];
        vector<vector<double>> dp(n + 1, vector<double>(n + 1));
        dp[0][0] = 0.5;
        for (int i = 1; i <= n; i++) {
            dp[i][0] = 0;
            dp[0][i] = 1;
        }
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                dp[i][j] = (dp[max(0, i - 4)][j] + dp[max(0, i - 3)][j - 1] + dp[max(0, i - 2)][max(0, j - 2)] + dp[i - 1][max(0, j - 3)]) / 4.0;

            }
        }
        return dp[n][n];
    }
};
// @lc code=end

